### ISSUE TITLE
[Training] Gradient Extraction in ONNX runtime training C++ 

### ISSUE THREAD

**Issue by zjc664656505** [microsoft/onnxruntime#16232]
### Describe the issue

Dear ONNX Community,

Currently, I'm work on a decentralized model parallel project using ONNX. Given a model, we first shard it to `n` sub-modules. Then, we are building p2p connection between devices in the computing cluster. 

In this setting, `device n` will send its sub-module forward output to `device n+1` and the `device n+1` will send its sub-module's gradient to `device n` for backpropagation.
+
As what we have seen so far, ONNX runtime training is achieved via the method `TrainStep` in the C++ API, which integrate the forward and backward into a single method.

With this method, we can get the sub-module's forward output directly. However, we cannot do backward without knowing the dependent sub-module's gradient on current sub-module, say now we want to do sub-module backward on `device n`, we need to have the gradient of the sub-module from `device n+1`. 

Therefore, I wish to ask whether there is a way to extract the gradient of a given model from the training session in the onnx runtime training C++ API? We are very new to onnx and wish to get some help.

Thanks!

### To reproduce

We will provide our code if it's requested.

### Urgency

This is urgent.

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.15

### PyTorch Version

2.1.0

### Execution Provider

Default CPU

### Execution Provider Library Version

1.18

**Comment by zjc664656505** [comment]
@baijumeswani @ashari4 @justinchuby @pengwa

**Comment by bitnom** [comment]
I found this researching the same as [#](https://github.com/microsoft/onnxruntime-training-examples/issues/143) . Just curious about the state/progress. This feature would be a big deal.
